### PR TITLE
filters: add rule filters

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve? 
continue merging placement rules.

### What is changed and how it works?
- labelConstraintFilter: selects stores satisfy the constraints.
- ruleFitFilter: make sure after replacing a peer, isolation level does not decrease.

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: server/schedule/filter/filters.go [hunk 1]
```diff
@@ -492,3 +494,78 @@ func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
...
+	fitter    RegionFitter
+	region    *core.RegionInfo
+	oldFit    *placement.RegionFit
+	IsLearner bool
+}
+
+// NewRuleFitFilter creates a filter that ensures that the region fitness won't decrease.
+func NewRuleFitFilter(scope string, fitter RegionFitter, region *core.RegionInfo, removePeer *metapb.Peer) *ruleFitFilter {
+	oldFit := fitter.FitRegion(region)
+	region = region.Clone(core.WithRemoveStorePeer(removePeer.GetStoreId()))
```
Comment (nolouch) [inline server/schedule/filter/filters.go]: How about record `origin region` and `replace store` to reduce once clone?

Comment (disksing) [reply server/schedule/filter/filters.go]: good idea.

File: server/schedule/filter/filters.go [hunk 2]
```diff
@@ -492,3 +493,76 @@ func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
...
+func (f labelConstraintFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
+	return !placement.MatchLabelConstraints(store, f.constraints)
+}
+
+// RegionFitter is the interface that can fit a region against placement rules.
+type RegionFitter interface {
+	FitRegion(*core.RegionInfo) *placement.RegionFit
+}
+
+type ruleFitFilter struct {
```
Comment (lhy1024) [inline server/schedule/filter/filters.go]: how about `ruleMatchFilter`?

File: server/schedule/filter/filters.go [hunk 3]
```diff
@@ -492,3 +493,76 @@ func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
...
+	scope   string
+	fitter  RegionFitter
+	region  *core.RegionInfo
+	oldFit  *placement.RegionFit
+	oldPeer uint64
+}
+
+// NewRuleFitFilter creates a filter that ensures that the region fitness won't decrease.
+func NewRuleFitFilter(scope string, fitter RegionFitter, region *core.RegionInfo, oldPeerID uint64) Filter {
```
Comment (lhy1024) [inline server/schedule/filter/filters.go]: I think we need more some comment about it

Comment (disksing) [reply server/schedule/filter/filters.go]: good idea.

File: server/schedule/filter/filters.go (hunk 2 above)
Comment (disksing) [reply server/schedule/filter/filters.go]: I think we should use `Fit` to keep it consistent with other places.

File: server/schedule/filter/filters.go [hunk 4]
```diff
@@ -492,3 +493,78 @@ func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
...
+type ruleFitFilter struct {
+	scope   string
+	fitter  RegionFitter
+	region  *core.RegionInfo
+	oldFit  *placement.RegionFit
+	oldPeer uint64
+}
+
+// NewRuleFitFilter creates a filter that ensures after replace a peer with new
+// one, the isolation level will not decrease. Its function is the same as
```
Comment (lhy1024) [inline server/schedule/filter/filters.go]:
```suggestion
// one, the isolation level of target store will not decrease. Its function is the same as
```

Comment (disksing) [reply server/schedule/filter/filters.go]: hmm, it is not the target store's isolation level. The isolation level belongs to a region. It is used to evaluate how a region's peers are isolated.

**Pull request merged**